The Enron fraud is a big, messy and totally fascinating story about corporate malfeasance of nearly every imaginable type. The Enron email and financial datasets are also big, messy treasure troves of information, which become much more useful once you know your way around them a bit. We’ve combined the email and finance data into a single dataset, which you’ll explore in this mini-project.

Getting started:

- Clone this git repository: https://github.com/udacity/ud120-projects
- Open the starter code: datasets_questions/explore_enron_data.py

The aggregated Enron email + financial dataset is stored in a dictionary, where each key in the dictionary is a person’s name and the value is a dictionary containing all the features of that person.

The email + finance (E+F) data dictionary is stored as a pickle file, which is a handy way to store and load python objects directly. Use datasets_questions/explore_enron_data.py to load the dataset.

How many data points (people) are in the dataset?

Starter code, "explore_enron_data.py":
```
""" 
    Starter code for exploring the Enron dataset (emails + finances);
    loads up the dataset (pickled dict of dicts).

    The dataset has the form:
    enron_data["LASTNAME FIRSTNAME MIDDLEINITIAL"] = { features_dict }

    {features_dict} is a dictionary of features associated with that person.
    You should explore features_dict as part of the mini-project,
    but here's an example to get you started:

    enron_data["SKILLING JEFFREY K"]["bonus"] = 5600000
    
"""
import pickle
enron_data = pickle.load(open('../final_project/final_project_dataset.pkl', 'r'))

```

In [1]:
import pickle

In [2]:
enron_data = pickle.load(open("../ud120-projects/final_project/final_project_dataset.pkl", "r"))

In [5]:
type(enron_data)

dict

#### How many data points (i.e. people) are in the Enron dataset?

In [6]:
len(enron_data.keys())

146

#### For each person, how many features are available?

In [7]:
type(enron_data.keys())

list

In [13]:
enron_data.keys()

['METTS MARK',
 'BAXTER JOHN C',
 'ELLIOTT STEVEN',
 'CORDES WILLIAM R',
 'HANNON KEVIN P',
 'MORDAUNT KRISTINA M',
 'MEYER ROCKFORD G',
 'MCMAHON JEFFREY',
 'HORTON STANLEY C',
 'PIPER GREGORY F',
 'HUMPHREY GENE E',
 'UMANOFF ADAM S',
 'BLACHMAN JEREMY M',
 'SUNDE MARTIN',
 'GIBBS DANA R',
 'LOWRY CHARLES P',
 'COLWELL WESLEY',
 'MULLER MARK S',
 'JACKSON CHARLENE R',
 'WESTFAHL RICHARD K',
 'WALTERS GARETH W',
 'WALLS JR ROBERT H',
 'KITCHEN LOUISE',
 'CHAN RONNIE',
 'BELFER ROBERT',
 'SHANKMAN JEFFREY A',
 'WODRASKA JOHN',
 'BERGSIEKER RICHARD P',
 'URQUHART JOHN A',
 'BIBI PHILIPPE A',
 'RIEKER PAULA H',
 'WHALEY DAVID A',
 'BECK SALLY W',
 'HAUG DAVID L',
 'ECHOLS JOHN B',
 'MENDELSOHN JOHN',
 'HICKERSON GARY J',
 'CLINE KENNETH W',
 'LEWIS RICHARD',
 'HAYES ROBERT E',
 'MCCARTY DANNY J',
 'KOPPER MICHAEL J',
 'LEFF DANIEL P',
 'LAVORATO JOHN J',
 'BERBERIAN DAVID',
 'DETMERING TIMOTHY J',
 'WAKEHAM JOHN',
 'POWERS WILLIAM',
 'GOLD JOSEPH',
 'BANNANTINE JAMES M',
 'DUNCAN JOHN H'

In [16]:
len(enron_data['RICE KENNETH D'])

21

#### The 'poi' feature records whether the person is a person of interest, according to our definition. How many POIs are there in the E+F dataset?

In other words, count the number of entries in the dictionary where
data[person_name]["poi"]==1

In [17]:
import pandas as pd
enron_data_df = pd.DataFrame.from_dict(enron_data)

In [18]:
enron_data_df.head()

ALLEN PHILLIP K BADUM JAMES P  \
bonus                              4175000           NaN   
deferral_payments                  2869717        178980   
deferred_income                   -3081055           NaN   
director_fees                          NaN           NaN   
email_address      phillip.allen@enron.com           NaN   

                           BANNANTINE JAMES M BAXTER JOHN C  \
bonus                                     NaN       1200000   
deferral_payments                         NaN       1295738   
deferred_income                         -5104      -1386055   
director_fees                             NaN           NaN   
email_address      james.bannantine@enron.com           NaN   

                        BAY FRANKLIN R BAZELIDES PHILIP J  \
bonus                           400000                NaN   
deferral_payments               260455             684694   
deferred_income                -201641                NaN   
director_fees                      NaN                NaN   
email_address      frank.bay@enron.com                NaN   

                           BECK SALLY W      BELDEN TIMOTHY N BELFER ROBERT  \
bonus                            700000               5249999           NaN   
deferral_payments                   NaN               2144013       -102500   
deferred_income                     NaN              -2334434           NaN   
director_fees                       NaN                   NaN          3285   
email_address      sally.beck@enron.com  tim.belden@enron.com           NaN   

                             BERBERIAN DAVID    ...     \
bonus                                    NaN    ...      
deferral_payments                        NaN    ...      
deferred_income                          NaN    ...      
director_fees                            NaN    ...      
email_address      david.berberian@enron.com    ...      

                             WASAFF GEORGE       WESTFAHL RICHARD K  \
bonus                               325000                      NaN   
deferral_payments                   831299                      NaN   
deferred_income                    -583325                   -10800   
director_fees                          NaN                      NaN   
email_address      george.wasaff@enron.com  dick.westfahl@enron.com   

                  WHALEY DAVID A      WHALLEY LAWRENCE G  \
bonus                        NaN                 3000000   
deferral_payments            NaN                     NaN   
deferred_income              NaN                     NaN   
director_fees                NaN                     NaN   
email_address                NaN  greg.whalley@enron.com   

                        WHITE JR THOMAS E WINOKUR JR. HERBERT S  \
bonus                              450000                   NaN   
deferral_payments                     NaN                   NaN   
deferred_income                       NaN                -25000   
director_fees                         NaN                108579   
email_address      thomas.white@enron.com                   NaN   

                             WODRASKA JOHN WROBEL BRUCE  \
bonus                                  NaN          NaN   
deferral_payments                      NaN          NaN   
deferred_income                        NaN          NaN   
director_fees                          NaN          NaN   
email_address      john.wodraska@enron.com          NaN   

                           YEAGER F SCOTT YEAP SOON  
bonus                                 NaN       NaN  
deferral_payments                     NaN       NaN  
deferred_income                       NaN       NaN  
director_fees                         NaN       NaN  
email_address      scott.yeager@enron.com       NaN  

[5 rows x 146 columns]

In [19]:
enron_data_dfT = enron_data_df.transpose()

In [21]:
enron_data_dfT.head()

bonus deferral_payments deferred_income director_fees  \
ALLEN PHILLIP K     4175000           2869717        -3081055           NaN   
BADUM JAMES P           NaN            178980             NaN           NaN   
BANNANTINE JAMES M      NaN               NaN           -5104           NaN   
BAXTER JOHN C       1200000           1295738        -1386055           NaN   
BAY FRANKLIN R       400000            260455         -201641           NaN   

                                 email_address exercised_stock_options  \
ALLEN PHILLIP K        phillip.allen@enron.com                 1729541   
BADUM JAMES P                              NaN                  257817   
BANNANTINE JAMES M  james.bannantine@enron.com                 4046157   
BAXTER JOHN C                              NaN                 6680544   
BAY FRANKLIN R             frank.bay@enron.com                     NaN   

                   expenses from_messages from_poi_to_this_person  \
ALLEN PHILLIP K       13868          2195                      47   
BADUM JAMES P          3486           NaN                     NaN   
BANNANTINE JAMES M    56301            29                      39   
BAXTER JOHN C         11200           NaN                     NaN   
BAY FRANKLIN R       129142           NaN                     NaN   

                   from_this_person_to_poi        ...         \
ALLEN PHILLIP K                         65        ...          
BADUM JAMES P                          NaN        ...          
BANNANTINE JAMES M                       0        ...          
BAXTER JOHN C                          NaN        ...          
BAY FRANKLIN R                         NaN        ...          

                   long_term_incentive    other    poi restricted_stock  \
ALLEN PHILLIP K                 304805      152  False           126027   
BADUM JAMES P                      NaN      NaN  False              NaN   
BANNANTINE JAMES M                 NaN   864523  False          1757552   
BAXTER JOHN C                  1586055  2660303  False          3942714   
BAY FRANKLIN R                     NaN       69  False           145796   

                   restricted_stock_deferred  salary shared_receipt_with_poi  \
ALLEN PHILLIP K                      -126027  201955                    1407   
BADUM JAMES P                            NaN     NaN                     NaN   
BANNANTINE JAMES M                   -560222     477                     465   
BAXTER JOHN C                            NaN  267102                     NaN   
BAY FRANKLIN R                        -82782  239671                     NaN   

                   to_messages total_payments total_stock_value  
ALLEN PHILLIP K           2902        4484442           1729541  
BADUM JAMES P              NaN         182466            257817  
BANNANTINE JAMES M         566         916197           5243487  
BAXTER JOHN C              NaN        5634343          10623258  
BAY FRANKLIN R             NaN         827696             63014  

[5 rows x 21 columns]

In [22]:
enron_data_dfT.poi.value_counts()

False    128
True      18
Name: poi, dtype: int64

#### How many POIs exist?

We compiled a list of all POI names (in ../final_project/poi_names.txt) and associated email addresses (in ../final_project/poi_email_addresses.py).

How many POI’s were there total? (Use the names file, not the email addresses, since many folks have more than one address and a few didn’t work for Enron, so we don’t have their emails.)

In [23]:
poi_names = pd.read_csv("../ud120-projects/final_project/poi_names.txt")

In [24]:
poi_names

http://usatoday30.usatoday.com/money/industries/energy/2005-12-28-enron-participants_x.htm
(y) Lay                                                   Kenneth                                        
(y) Skilling                                              Jeffrey                                        
(n) Howard                                                  Kevin                                        
(n) Krautz                                                Michael                                        
(n) Yeager                                                  Scott                                        
(n) Hirko                                                  Joseph                                        
(n) Shelby                                                    Rex                                        
(n) Bermingham                                              David                                        
(n) Darby                                                   Giles                                        
(n) Mulgrew                                                  Gary                                        
(n) Bayley                                                 Daniel                                        
(n) Brown                                                   James                                        
(n) Furst                                                  Robert                                        
(n) Fuhs                                                  William                                        
(n) Causey                                                Richard                                        
(n) Calger                                            Christopher                                        
(n) DeSpain                                               Timothy                                        
(n) Hannon                                                  Kevin                                        
(n) Koenig                                                   Mark                                        
(y) Forney                                                   John                                        
(n) Rice                                                  Kenneth                                        
(n) Rieker                                                  Paula                                        
(n) Fastow                                                    Lea                                        
(n) Fastow                                                 Andrew                                        
(y) Delainey                                                David                                        
(n) Glisan                                                    Ben                                        
(n) Richter                                               Jeffrey                                        
(n) Lawyer                                                  Larry                                        
(n) Belden                                                Timothy                                        
(n) Kopper                                                Michael                                        
(n) Duncan                                                  David                                        
(n) Bowen                                                 Raymond                                        
(n) Colwell                                                Wesley                                        
(n) Boyle                                                     Dan                                        
(n) Loehr                                             Christopher

In [25]:
len(poi_names)

35

#### Problems with incomplete data

As you can see, we have many of the POIs in our E+F dataset, but not all of them. Why is that a potential problem?

We will return to this later to explain how a POI could end up not being in the Enron E+F dataset, so you fully understand the issue before moving on.


Incomplete data = not enough data for training

#### Query the Dataset 1

Like any dict of dicts, individual people/features can be accessed like so:

enron_data["LASTNAME FIRSTNAME"]["feature_name"]
or, sometimes 
enron_data["LASTNAME FIRSTNAME MIDDLEINITIAL"]["feature_name"]

What is the total value of the stock belonging to James Prentice?

In [28]:
enron_data["PRENTICE JAMES"]['total_stock_value']

1095040

#### Query the Dataset 2

Like any dict of dicts, individual people/features can be accessed like so:

enron_data["LASTNAME FIRSTNAME"]["feature_name"]

How many email messages do we have from Wesley Colwell to persons of interest?

In [29]:
enron_data["COLWELL WESLEY"]["from_this_person_to_poi"]

11

#### Query the Dataset 3

Like any dict of dicts, individual people/features can be accessed like so:

```
enron_data["LASTNAME FIRSTNAME"]["feature_name"]

or

enron_data["LASTNAME FIRSTNAME MIDDLEINITIAL"]["feature_name"]
```

What’s the value of stock options exercised by Jeffrey K Skilling?

In [30]:
enron_data["SKILLING JEFFREY K"]["exercised_stock_options"]

19250000

#### Research the Enron Fraud

In the coming lessons, we’ll talk about how the best features are often motivated by our human understanding of the problem at hand. In this case, that means knowing a little about the story of the Enron fraud.

If you have an hour and a half to spare, “Enron: The Smartest Guys in the Room” is a documentary that gives an amazing overview of the story. Alternatively, there are plenty of archival newspaper stories that chronicle the rise and fall of Enron.

Which of these schemes was Enron NOT involved in?

- selling assets to shell companies at the end of each month, and buying them back at the beginning of the next month to hide accounting losses
- causing electrical grid failures in California
- X illegally obtained a government report that enabled them to corner the market on frozen concentrated orange juice futures
- X conspiring to give a Saudi prince expedited American citizenship
- a plan in collaboration with Blockbuster movies to stream movies over the internet



#### Enron CEO

Who was the CEO of Enron during most of the time that fraud was being perpetrated?

- Jeff Skilling

#### Enron Chairman

Who was chariman of the Enron Board of directors?

- Ken Lay

#### Enron CFO

Who was CFO of Enron during most of the time that fraud was going on?

- Andrew Fastow

#### Follow the Money

Of these three individuals, who took home the most money (largest value of "total_payments" feature)?

In [50]:
enron_data.keys()

['METTS MARK',
 'BAXTER JOHN C',
 'ELLIOTT STEVEN',
 'CORDES WILLIAM R',
 'HANNON KEVIN P',
 'MORDAUNT KRISTINA M',
 'MEYER ROCKFORD G',
 'MCMAHON JEFFREY',
 'HORTON STANLEY C',
 'PIPER GREGORY F',
 'HUMPHREY GENE E',
 'UMANOFF ADAM S',
 'BLACHMAN JEREMY M',
 'SUNDE MARTIN',
 'GIBBS DANA R',
 'LOWRY CHARLES P',
 'COLWELL WESLEY',
 'MULLER MARK S',
 'JACKSON CHARLENE R',
 'WESTFAHL RICHARD K',
 'WALTERS GARETH W',
 'WALLS JR ROBERT H',
 'KITCHEN LOUISE',
 'CHAN RONNIE',
 'BELFER ROBERT',
 'SHANKMAN JEFFREY A',
 'WODRASKA JOHN',
 'BERGSIEKER RICHARD P',
 'URQUHART JOHN A',
 'BIBI PHILIPPE A',
 'RIEKER PAULA H',
 'WHALEY DAVID A',
 'BECK SALLY W',
 'HAUG DAVID L',
 'ECHOLS JOHN B',
 'MENDELSOHN JOHN',
 'HICKERSON GARY J',
 'CLINE KENNETH W',
 'LEWIS RICHARD',
 'HAYES ROBERT E',
 'MCCARTY DANNY J',
 'KOPPER MICHAEL J',
 'LEFF DANIEL P',
 'LAVORATO JOHN J',
 'BERBERIAN DAVID',
 'DETMERING TIMOTHY J',
 'WAKEHAM JOHN',
 'POWERS WILLIAM',
 'GOLD JOSEPH',
 'BANNANTINE JAMES M',
 'DUNCAN JOHN H'

In [51]:
enron_data["LAY KENNETH L"]["total_payments"]

103559793

In [53]:
enron_data["FASTOW ANDREW S"]["total_payments"]

2424083

In [54]:
enron_data["SKILLING JEFFREY K"]["total_payments"]

8682716

#### Unfilled Features

For nearly every person in the dataset, not every feature has a value. How is it denoted when a feature doesn't have a well-defined value?

- NaN

#### Dealing with Unfilled Features

How many folks in the dataset have a quantified salary? What about a known email address?

In [55]:
len(enron_data_dfT[enron_data_dfT.salary == 'NaN'])

51

In [56]:
len(enron_data_dfT) - len(enron_data_dfT[enron_data_dfT.salary == 'NaN'])

95

In [57]:
len(enron_data_dfT) - len(enron_data_dfT[enron_data_dfT.email_address == 'NaN'])

111

#### Dict-to-array conversion

A python dictionary can’t be read directly into an sklearn classification or regression algorithm; instead, it needs a numpy array or a list of lists (each element of the list (itself a list) is a data point, and the elements of the smaller list are the features of that point).

We’ve written some helper functions (featureFormat() and targetFeatureSplit() in tools/feature_format.py) that can take a list of feature names and the data dictionary, and return a numpy array.

In the case when a feature does not have a value for a particular person, this function will also replace the feature value with 0 (zero).

In [58]:
""" 
    A general tool for converting data from the
    dictionary format to an (n x k) python list that's 
    ready for training an sklearn algorithm
    
    n--no. of key-value pairs in dictonary
    k--no. of features being extracted
    
    dictionary keys are names of persons in dataset
    dictionary values are dictionaries, where each
        key-value pair in the dict is the name
        of a feature, and its value for that person
    
    In addition to converting a dictionary to a numpy 
    array, you may want to separate the labels from the
    features--this is what targetFeatureSplit is for
    so, if you want to have the poi label as the target,
    and the features you want to use are the person's
    salary and bonus, here's what you would do:
    
    feature_list = ["poi", "salary", "bonus"] 
    data_array = featureFormat( data_dictionary, feature_list )
    label, features = targetFeatureSplit(data_array)
    
    the line above (targetFeatureSplit) assumes that the
    label is the _first_ item in feature_list--very important
    that poi is listed first!
"""

import numpy as np

In [59]:
def featureFormat( dictionary, features, remove_NaN=True, remove_all_zeroes=True, remove_any_zeroes=False, sort_keys = False):
    """ convert dictionary to numpy array of features
        remove_NaN = True will convert "NaN" string to 0.0
        remove_all_zeroes = True will omit any data points for which
            all the features you seek are 0.0
        remove_any_zeroes = True will omit any data points for which
            any of the features you seek are 0.0
        sort_keys = True sorts keys by alphabetical order. Setting the value as
            a string opens the corresponding pickle file with a preset key
            order (this is used for Python 3 compatibility, and sort_keys
            should be left as False for the course mini-projects).
        NOTE: first feature is assumed to be 'poi' and is not checked for
            removal for zero or missing values.
    """


    return_list = []

    # Key order - first branch is for Python 3 compatibility on mini-projects,
    # second branch is for compatibility on final project.
    if isinstance(sort_keys, str):
        import pickle
        keys = pickle.load(open(sort_keys, "rb"))
    elif sort_keys:
        keys = sorted(dictionary.keys())
    else:
        keys = dictionary.keys()

    for key in keys:
        tmp_list = []
        for feature in features:
            try:
                dictionary[key][feature]
            except KeyError:
                print "error: key ", feature, " not present"
                return
            value = dictionary[key][feature]
            if value=="NaN" and remove_NaN:
                value = 0
            tmp_list.append( float(value) )

        # Logic for deciding whether or not to add the data point.
        append = True
        # exclude 'poi' class as criteria.
        if features[0] == 'poi':
            test_list = tmp_list[1:]
        else:
            test_list = tmp_list
        ### if all features are zero and you want to remove
        ### data points that are all zero, do that here
        if remove_all_zeroes:
            append = False
            for item in test_list:
                if item != 0 and item != "NaN":
                    append = True
                    break
        ### if any features for a given data point are zero
        ### and you want to remove data points with any zeroes,
        ### handle that here
        if remove_any_zeroes:
            if 0 in test_list or "NaN" in test_list:
                append = False
        ### Append the data point if flagged for addition.
        if append:
            return_list.append( np.array(tmp_list) )

    return np.array(return_list)

In [60]:
def targetFeatureSplit( data ):
    """ 
        given a numpy array like the one returned from
        featureFormat, separate out the first feature
        and put it into its own list (this should be the 
        quantity you want to predict)
        return targets and features as separate lists
        (sklearn can generally handle both lists and numpy arrays as 
        input formats when training/predicting)
    """

    target = []
    features = []
    for item in data:
        target.append( item[0] )
        features.append( item[1:] )

    return target, features

#### Missing POIs 1

As you saw a little while ago, not every POI has an entry in the dataset (e.g. Michael Krautz). That’s because the dataset was created using the financial data you can find in final_project/enron61702insiderpay.pdf, which is missing some POI’s (those absences propagated through to the final dataset). On the other hand, for many of these “missing” POI’s, we do have emails.

While it would be straightforward to add these POI’s and their email information to the E+F dataset, and just put “NaN” for their financial information, this could introduce a subtle problem. You will walk through that here.

How many people in the E+F dataset (as it currently exists) have “NaN” for their total payments? What percentage of people in the dataset as a whole is this?

In [72]:
#enron_data_dfT
enron_data_dfT["total_payments"].value_counts(normalize=True)

NaN          0.143836
1279         0.006849
102500       0.006849
1208649      0.006849
762135       0.006849
213071       0.006849
1101393      0.006849
1934359      0.006849
1321557      0.006849
8682716      0.006849
4484442      0.006849
1099100      0.006849
911453       0.006849
618850       0.006849
1545059      0.006849
1086821      0.006849
1747522      0.006849
2151         0.006849
2047593      0.006849
1318763      0.006849
969068       0.006849
362096       0.006849
875889       0.006849
87410        0.006849
2014835      0.006849
2101364      0.006849
477557       0.006849
1092663      0.006849
3859065      0.006849
228474       0.006849
               ...   
319941       0.006849
87492        0.006849
309886585    0.006849
49288        0.006849
189583       0.006849
1204583      0.006849
148          0.006849
2368151      0.006849
1034395      0.006849
2146973      0.006849
50591        0.006849
1410464      0.006849
812194       0.006849
1848227      0.006849
860136    

Answer: 21 people out of 146, 14.4%

#### Missing POIs 2

How many POIs in the E+F dataset have "NaN" for their total payments? What percentage of POIs as a whole is this?

In [81]:
#enron_data_dfT.keys()
#enron_data_dfT.poi
poi_EF = enron_data_dfT[enron_data_dfT.poi == True]

In [84]:
poi_EF.total_payments.value_counts()

5501630      1
1587421      1
1639297      1
2652612      1
1099100      1
1490344      1
288682       1
4747979      1
360300       1
2003885      1
103559793    1
2424083      1
91093        1
1868758      1
8682716      1
505050       1
1272284      1
2669589      1
Name: total_payments, dtype: int64

Answer => 0

#### Missing POIs 3

If a machine learning algorithm were to use total_payments as a feature, would you expect it to associate a "NaN" value with POIs or non-POIs?

Answer: non-POIs

#### Misiong POIs 4

If you added in, say, 10 more data points which were all POI’s, and put “NaN” for the total payments for those folks, the numbers you just calculated would change.
What is the new number of people of the dataset? 

156

What is the new number of folks with “NaN” for total payments?

31

This makes for 20% of them with a "NaN" overall.


#### Missing POIs 5

What is the new number of POIs in the dataset?

28

What is the new number of POIs with NaN for total_payments?

10

That's 36% of the POI's who have "NaN" for total_payments, a big jump from before.

In [86]:
len(poi_EF) + 10

28

#### Missing POIs 6

Once the new data points are added, do you think a supervised clasificaiton algorithm might interpret "NaN" for total_payments as a clue that someone is a POI?

Yes

#### Mixing Data Sources

Adding in the new POI’s in this example, none of whom we have financial information for, has introduced a subtle problem, that our lack of financial information about them can be picked up by an algorithm as a clue that they’re POIs. Another way to think about this is that there’s now a difference in how we generated the data for our two classes--non-POIs all come from the financial spreadsheet, while many POIs get added in by hand afterwards. That difference can trick us into thinking we have better performance than we do--suppose you use your POI detector to decide whether a new, unseen person is a POI, and that person isn’t on the spreadsheet. Then all their financial data would contain “NaN” but the person is very likely not a POI (there are many more non-POIs than POIs in the world, and even at Enron)--you’d be likely to accidentally identify them as a POI, though!

This goes to say that, when generating or augmenting a dataset, you should be exceptionally careful if your data are coming from different sources for different classes. It can easily lead to the type of bias or mistake that we showed here. There are ways to deal with this, for example, you wouldn’t have to worry about this problem if you used only email data--in that case, discrepancies in the financial data wouldn’t matter because financial features aren’t being used. There are also more sophisticated ways of estimating how much of an effect these biases can have on your final answer; those are beyond the scope of this course.

For now, the takeaway message is to be very careful about introducing features that come from different sources depending on the class! It’s a classic way to accidentally introduce biases and mistakes.
